File: eel-core/src/main/kotlin/io/eels/component/orc/OrcSink.scala
package io.eels.component.orc

import io.eels.{Schema, Sink, SinkWriter}
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path
import org.apache.hadoop.hive.ql.io.orc.OrcFile
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory
import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory

case class OrcSink(path: Path) extends Sink with StrictLogging {
  override def writer(schema: Schema): SinkWriter = new OrcSinkWriter(schema, path)
}

class OrcSinkWriter(schema: Schema, path: Path) extends SinkWriter with StrictLogging {
  self =>
  logger.debug(s"Creating orc writer $schema")

  val inspector = ObjectInspectorFactory.getStandardListObjectInspector(
    PrimitiveObjectInspectorFactory.javaStringObjectInspector
  )
  val writer = OrcFile.createWriter(path, OrcFile.writerOptions(new Configuration).inspector(inspector))
  writer.addRow(schema.columnNames.toArray)

  override def write(row: InternalRow): Unit = {
    self.synchronized {
      writer.addRow(row.toArray)
    }
  }

  override def close(): Unit = writer.close()
}

File: eel-core/src/main/kotlin/io/eels/component/orc/OrcSink.kt
package io.eels.component.orc

import io.eels.util.Logging
import io.eels.Row
import io.eels.schema.Schema
import io.eels.Sink
import io.eels.SinkWriter
import org.apache.hadoop.hive.ql.io.orc.OrcFile
import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory
import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory

import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path
import org.apache.hadoop.hive.ql.io.orc.Writer
import org.apache.hadoop.hive.serde2.objectinspector.StandardListObjectInspector

class OrcSink(val path: Path) : Sink, Logging {
  override fun writer(schema: Schema): SinkWriter = OrcSinkWriter(schema, path)
}

class OrcSinkWriter(schema: Schema, path: Path) : SinkWriter, Logging {
  init {
    logger.debug("Creating orc writer $schema")
  }

  val inspector: StandardListObjectInspector = ObjectInspectorFactory.getStandardListObjectInspector(
    PrimitiveObjectInspectorFactory.javaStringObjectInspector
  )

  val writer: Writer = OrcFile.createWriter(path, OrcFile.writerOptions(Configuration()).inspector(inspector)).apply {
    this.addRow(schema.fieldNames().toTypedArray())
  }

  override fun write(row: Row): Unit {
    synchronized(this) {
      writer.addRow(row.values)
    }
  }

  override fun close(): Unit = writer.close()
}
